### PR TITLE
Add validation for api/import

### PR DESCRIPTION
**Pull request by jimchamp**
Work towards #5833 

Adds required field validations to the `api/import` endpoint.  Two types of validation occur for each required field.  We first check for the field's existence, then we check that the value is formatted properly (correct type, non-empty strings and lists, etc).

If validation fails, a list of invalid fields is returned and the data will not be imported.

### Technical

This schema was used to determine what fields are required: https://github.com/internetarchive/openlibrary-client/blob/master/olclient/schemata/import.schema.json

Validation occurs as the import_edition_builder is initialized, and before the `add_book.load()` call.  It is important to validate before the `load()` call, as errors raised by `load()` can sometimes terminate our automated imports.

Validation is fail-fast.  If the given data contains both missing required fields and invalid required values, the error message will only contain the missing required fields.

### Testing

I tested this locally by POSTing valid and invalid import JSON strings to `http://localhost:8080/api/import`.  Similar tools can also be used.  A session needs to be created before POSTing.

### CODE REVIEW

**Comment by cclauss** [comment]
I really think we should use Pydantic to automate this.  I created an example of how it works...
 https://github.com/cclauss/pydantic-books-example#pydantic-books-example

**Comment by jimchamp** [comment]
I think that that is a good idea.  Will update this code tomorrow. Thanks!

**Comment by cdrini** [comment]
Note:

- We should make sure to use the openlibrary-client schema files for validation
- Probably move those schema in to this repo; and then we need to ensure they stay in sync; either find a way to fetch/cache from this repo, or maybe better, leave a copy, but have a unit test that ensure it's up-to-date.

File: openlibrary/plugins/importapi/import_validator.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        If any required fields are missing, a RequiredFieldError containing
+        the missing keys is raised and this method terminates without
+        further validation.
+
+        Returns True if the import object is valid.
+        """
+
+        missing_keys = [key for key in REQUIRED_KEYS if key not in data]
+        if len(missing_keys):
+            raise RequiredFieldError(missing_keys)
```
[the remark this answers is not in the document]
Comment (jimchamp) [reply openlibrary/plugins/importapi/import_validator.py]: Great approach!  Resolving, as this code no longer exists.

**Comment by jimchamp** [comment]
Just pushed update that uses Pydantic for import object validation.  I like how Pydantic returns all validation errors at once.  We use `jsonschema` for validation in openlibrary-client, and it does not aggregate validation errors.

The schemata have not yet been imported to this repo.  If we want to use them for validation, we will have to update the import schema by either:
1. Adding an `identifiers` field, which takes a list of objects with a single entry, or
2. Change the following value to true:
https://github.com/internetarchive/openlibrary-client/blob/99e600136f033a034690f947e3f09d930455c2c9/olclient/schemata/import.schema.json#L12

Option two may be worse than our existing limited validation.

There's an additional question of how to validate once the schemata is in this repo.  Should we use `jsonschema` or Pydantic to validate? [This tool](https://github.com/koxudaxi/datamodel-code-generator/) can be used to generate Pydantic classes from our schemata.  This would give us Pydantic's wonderful error reporting with the exact import object specifications.  Sample generated file can be found [here](https://gist.github.com/jimchamp/71fb38e124f671efec38a890c2ccd025).

Let me know what you think, @cdrini and @cclauss.

**Comment by cclauss** [comment]
> I like how Pydantic returns all validation errors at once.

That plus the quality/precision of the error messages are of great value.

https://github.com/cclauss/pydantic-books-example/blob/main/src/Books/work.py might be of interest because it does `Work(**requests.get().json())` to build an Open Library work from our own json data.  It then lazy loads all authors and languages as a further demonstration of the nested data validation that we could do on external APIs.

**Review by cclauss — APPROVED** [review]
LGTM.  Thanks for the time on Zoom explaining how you test this and how it deals with extraneous data.  This is a nice step in the direction of getting cleaner data on imports.

**Review by cdrini — APPROVED** [review]
Tested:

- ✅ Missing fields
```js
await fetch('/api/import', {
    method: "POST",
    body: JSON.stringify({ title: "Beowulf" })
})
---
{"success": false, "error_code": "invalid-value", "error": "4 validation errors for Book: source_records:   field required (type=value_error.missing): authors:   field required (type=value_error.missing): publishers:   field required (type=value_error.missing): publish_date:   field required (type=value_error.missing)"}
```

- ✅ Extra fields cause no issue (for now!)

```js
await fetch('/api/import', {
    method: "POST",
    body: JSON.stringify({
        title: "Beowulf",
        source_records: ["key:value"],
        helter_skelter: 'in a summer swelter',
        authors: [{"name": "Tom Robbins"}],
        publishers: ["Harper Collins", "OpenStax"],
        publish_date: "December 2018"
    })
})
```

- ✅ Empty required fields not allowed

```js
await fetch('/api/import', {
    method: "POST",
    body: JSON.stringify({
        title: "",
        source_records: ["key:value"],
        authors: [{"name": "Tom Robbins"}],
        publishers: ["Harper Collins", "OpenStax"],
        publish_date: "December 2018"
    })
})
---
{"success": false, "error_code": "invalid-value", "error": "1 validation error for Book: title:   Field must have a non-empty string value (type=value_error)"}
```

- ✅ Matches if same thing is imported twice


Lgtm! Now that we have the DRY-ing up clearly outlined in the next steps.

**Pull request merged**